src/pool/srv_util.c

```diff
@@ -202,91 +202,625 @@ ds_pool_transfer_map_buf(struct pool_buf *map_buf, uint32_t map_version,
+	D_ASSERT(rc == 0);
+
+	pool_map_decref(map);
+}
+
+void
+ds_pool_test_plan_svc_reconfs(void)
+{
+	d_rank_t		self = 0;
+	d_rank_list_t	       *to_add;
+	d_rank_list_t	       *to_remove;
+
+#define call_testu_plan_svc_reconfs								\
+	testu_plan_svc_reconfs(svc_rf, ranks, ARRAY_SIZE(ranks), down_ranks,			\
+			       ARRAY_SIZE(down_ranks), replicas_ranks,				\
+			       ARRAY_SIZE(replicas_ranks), self, &to_add, &to_remove);
+
+#define call_d_rank_list_free									\
+	d_rank_list_free(to_add);								\
+	d_rank_list_free(to_remove);
+
+	/* A happy PS does not want any changes. */
+	{
+		int		svc_rf = 2;
+		d_rank_t	ranks[] = {0, 1, 2, 3, 4, 5, 6, 7};
+		d_rank_t	down_ranks[] = {};
+		d_rank_t	replicas_ranks[] = {0, 1, 2, 3, 4};
+
+		call_testu_plan_svc_reconfs
+
+		D_ASSERT(to_add->rl_nr == 0);
+		D_ASSERT(to_remove->rl_nr == 0);
+
+		call_d_rank_list_free
+	}
+
+	/* One lonely replica. */
+	{
+		int		svc_rf = 0;
+		d_rank_t	ranks[] = {0};
+		d_rank_t	down_ranks[] = {};
+		d_rank_t	replicas_ranks[] = {0};
+
+		call_testu_plan_svc_reconfs
+
+		D_ASSERT(to_add->rl_nr == 0);
+		D_ASSERT(to_remove->rl_nr == 0);
+
+		call_d_rank_list_free
+	}
+
+	/* A PS tries to achieve the RF. */
+	{
+		int		svc_rf = 1;
+		d_rank_t	ranks[] = {0, 1};
+		d_rank_t	down_ranks[] = {};
+		d_rank_t	replicas_ranks[] = {0};
+		d_rank_t	expected_to_add[] = {1};
+
+		call_testu_plan_svc_reconfs
+
+		D_ASSERT(testu_rank_sets_identical(to_add, expected_to_add,
+						   ARRAY_SIZE(expected_to_add)));
+		D_ASSERT(to_remove->rl_nr == 0);
+
+		call_d_rank_list_free
+	}
+
+	/* A PS resists the temptation of down ranks. */
+	{
+		int		svc_rf = 1;
+		d_rank_t	ranks[] = {0, 1, 2};
+		d_rank_t	down_ranks[] = {2};
+		d_rank_t	replicas_ranks[] = {0};
+		d_rank_t	expected_to_add[] = {1};
+
+		call_testu_plan_svc_reconfs
+
+		D_ASSERT(testu_rank_sets_identical(to_add, expected_to_add,
+						   ARRAY_SIZE(expected_to_add)));
+		D_ASSERT(to_remove->rl_nr == 0);
+
+		call_d_rank_list_free
+	}
+
+	/* A PS successfully achieves the RF. */
+	{
+		int		svc_rf = 1;
+		d_rank_t	ranks[] = {0, 1, 2};
+		d_rank_t	down_ranks[] = {};
+		d_rank_t	replicas_ranks[] = {0};
+		d_rank_t	expected_to_add[] = {1, 2};
+
+		call_testu_plan_svc_reconfs
+
+		D_ASSERT(testu_rank_sets_identical(to_add, expected_to_add,
+						   ARRAY_SIZE(expected_to_add)));
+		D_ASSERT(to_remove->rl_nr == 0);
+
+		call_d_rank_list_free
+	}
+
+	/* A PS holds its ground when there's no replacement. */
+	{
+		int		svc_rf = 1;
+		d_rank_t	ranks[] = {0, 1, 2};
+		d_rank_t	down_ranks[] = {2};
+		d_rank_t	replicas_ranks[] = {0, 1, 2};
+
+		call_testu_plan_svc_reconfs
+
+		D_ASSERT(to_add->rl_nr == 0);
+		D_ASSERT(to_remove->rl_nr == 0);
+
+		call_d_rank_list_free
+	}
+
+	/* A PS replaces one down rank. */
+	{
+		int		svc_rf = 1;
+		d_rank_t	ranks[] = {0, 1, 2, 3, 4};
+		d_rank_t	down_ranks[] = {2};
+		d_rank_t	replicas_ranks[] = {0, 1, 2};
+		d_rank_t	expected_to_add_candidates[] = {3, 4};
```

Comment:
No changes suggested, just confirming understanding. Intuitively I would think for expected_to_add_candidates[] to be = {3}, and to use testu_rank_sets_identical() below.
However the test expects exactly one of these ranks to replace the down rank 2. I guess this is done to not expect a particular ordering of the domains/ranks in the pool map?

Reply:
Yes. Another consideration is that we'll probably make the replica placement a bit more random in the future (Mike has already made such a contribution that hasn't been enabled). Will add a general comment above if this patch needs another refresh.